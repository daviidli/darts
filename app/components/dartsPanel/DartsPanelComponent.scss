.totalsContainer {
	display: flex;
	align-items: center;
	justify-content: center;
	height: 70%;

	.totals {
		font-size: 14em;
		font-weight: 900;
		text-shadow: 0.06em 0.06em #e49708;
		text-align: center;
	}

	.dartsContainer {
		height: 30%;

		.dart {
			font-size: 3em;
			font-weight: 700;
			text-align: center;
			padding: 0.5em;
			width: 20em;

			&.active {
				background-color: rgba(51, 169, 255, 0.4);
				border-radius: 10em;
			}
		}
	}
}
