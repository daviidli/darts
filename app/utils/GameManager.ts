/* eslint-disable no-await-in-loop */
import DartBoard from './dartBoard';
import store from '../store/store';
import {
	setCurrentRound,
	setCurrentPlayer,
	initializedRound,
	setWinner,
	setWaiting,
	setMaxDarts,
	setMaxRounds,
	setCurrentDart,
	setTotals
} from '../actions/gameActions';

abstract class GameManager {
	private dartboard: DartBoard;
	private running: boolean;
	private missedDart: boolean;
	private dartsCleared: boolean;

	constructor() {
		const state = store.getState();

		if (!state.serialPort.length) {
			throw new Error('No serial port selected.');
		}

		this.dartboard = new DartBoard(state.serialPort);
		this.running = false;
		this.missedDart = false;
		this.dartsCleared = false;
	}

	protected abstract initialize(): void;
	protected abstract async turn(): Promise<void>;

	protected async run() {
		const {
			maxRounds,
			maxDarts,
			players,
			clearDartsWaitTime
		} = store.getState();

		const roundsArr = this.createStrArray(
			players.length,
			maxRounds,
			maxDarts
		);
		store.dispatch(initializedRound(roundsArr));

		this.running = true;
		for (let r = 0; r < maxRounds && this.running; r++) {
			store.dispatch(setCurrentRound(r));
			for (let p = 0; p < players.length && this.running; p++) {
				store.dispatch(setCurrentPlayer(p));
				await this.turn();
				await this.waitForClearingDarts(clearDartsWaitTime);
			}
		}
	}

	protected async getDart(): Promise<string> {
		let currentDart;
		while (true) {
			try {
				currentDart = await this.dartboard.getResults(1, 500);
			} catch (e) {
				// continue regardless of error
				// todo: show error notification
			}

			if (this.missedDart) {
				this.missedDart = false;
				return 'miss';
			}

			if (currentDart) {
				return currentDart[0];
			}
		}
	}

	public start() {
		this.initialize();
		this.run();
	}

	public stop() {
		this.running = false;
		this.dartboard.close();
		store.dispatch(setMaxDarts(0));
		store.dispatch(setMaxRounds(0));
		store.dispatch(setCurrentPlayer(0));
		store.dispatch(setCurrentDart(0));
		store.dispatch(setCurrentRound(0));
		store.dispatch(initializedRound([]));
		store.dispatch(setTotals([]));
	}

	public setWinner(index: number) {
		this.stop();
		store.dispatch(setWinner(index));
	}

	public miss() {
		this.missedDart = true;
	}

	public getRunning() {
		return this.running;
	}

	public clearedDarts() {
		this.dartsCleared = true;
	}

	private async waitForClearingDarts(timeout = 0) {
		this.dartsCleared = false;
		store.dispatch(setWaiting(true));
		let timeouts = 0;

		while (true) {
			await this.timeout(1000);
			timeouts += 1;
			if (timeouts === timeout) {
				store.dispatch(setWaiting(false));
				return;
			}
			if (this.dartsCleared) {
				this.dartsCleared = false;
				store.dispatch(setWaiting(false));
				return;
			}
		}
	}

	private timeout(ms: number): Promise<void> {
		return new Promise(resolve => setTimeout(resolve, ms));
	}

	private createStrArray(
		iMax: number,
		jMax: number,
		kMax: number
	): string[][][] {
		const initialRounds: string[][][] = [];
		for (let i = 0; i < iMax; i++) {
			const arr = [];
			for (let j = 0; j < jMax; j++) {
				arr.push(new Array(kMax).fill(''));
			}
			initialRounds.push(arr);
		}
		return initialRounds;
	}
}

export default GameManager;
