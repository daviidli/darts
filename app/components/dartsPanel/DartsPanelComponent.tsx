/* eslint-disable react/no-array-index-key */
import React from 'react';
import { Grid, Typography } from '@material-ui/core';
import clsx from 'clsx';
import { GiDart } from 'react-icons/gi';
import GameManager from '../../utils/GameManager';
import styles from './DartsPanelComponent.scss';
import { mappingToString } from '../../utils/dartUtils';
import DartText from '../dartText/DartText';

export type Props = {
	totals: any[];
	rounds: string[][][];
	currentPlayer: number;
	currentRound: number;
	currentDart: number;
	gameManager: GameManager;
};

const DartsPanelComponent = (props: Props) => {
	const { totals, rounds, currentPlayer, currentRound, currentDart } = props;

	return (
		<>
			<div className={styles.totalsContainer}>
				<Typography variant="h2" className={styles.totals}>
					{totals[currentPlayer]}
				</Typography>
			</div>
			<Grid
				container
				direction="row"
				justify="center"
				alignItems="center"
				className={styles.dartsContainer}
			>
				{rounds[currentPlayer][currentRound].map((dart, i) => {
					let content;

					if (dart === '') {
						content = <GiDart />;
					} else {
						content = <DartText value={mappingToString(dart)} />;
					}

					return (
						<Grid
							item
							xs
							key={`round-${i}`}
							className={clsx(styles.dart, {
								[styles.active]: i === currentDart
							})}
						>
							{content}
						</Grid>
					);
				})}
			</Grid>
		</>
	);
};

export default DartsPanelComponent;
