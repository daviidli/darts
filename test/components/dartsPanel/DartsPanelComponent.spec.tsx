import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import DartsPanelComponent, {
	Props
} from '../../../app/components/dartsPanel/DartsPanelComponent';
import GameManager from '../../../app/utils/GameManager';

Enzyme.configure({ adapter: new Adapter() });

const initialProps: Props = {
	totals: [0],
	rounds: [[['1', '']]],
	currentPlayer: 0,
	currentRound: 0,
	currentDart: 0,
	gameManager: {} as GameManager
};

const setup = (props: Props = initialProps) => {
	const wrapper = shallow(<DartsPanelComponent {...props} />);
	return { wrapper };
};

describe('DartsPanel component', () => {
	it('should match snapshot', () => {
		const { wrapper } = setup();
		expect(toJson(wrapper)).toMatchSnapshot();
	});
});
