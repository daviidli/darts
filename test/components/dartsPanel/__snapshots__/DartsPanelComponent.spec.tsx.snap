// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DartsPanel component should match snapshot 1`] = `
<Fragment>
  <div
    className="totalsContainer"
  >
    <WithStyles(ForwardRef(Typography))
      className="totals"
      variant="h2"
    >
      0
    </WithStyles(ForwardRef(Typography))>
  </div>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    className="dartsContainer"
    container={true}
    direction="row"
    justify="center"
  >
    <WithStyles(ForwardRef(Grid))
      className="dart active"
      item={true}
      key="round-0"
      xs={true}
    >
      <DartText
        value={
          Object {
            "modifier": "",
            "value": "1",
          }
        }
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="dart"
      item={true}
      key="round-1"
      xs={true}
    >
      <GiDart />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</Fragment>
`;
